lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,353 @@
+	// an operator. Since we have not encountered a newline, that scenario is no longer possible and the literal value
+	// is used as part of a word, so we delimit the operator at this point.
+	if (state_data.previous == State::Operator) {
+		delimitOperator();
+	}
+
+	state_data.word.push_back('\\');
+	state_data.word.push_back(ch);
+	chars.consume();
+	return State::Word;
+}
+
+Lexer::State Lexer::singleQuoteStartState() {
+	char ch = chars.consume();
+	D_ASSERT(isSingleQuote(ch));
+	state_data.word.push_back(ch);
+	return State::SingleQuote;
+}
+
+Lexer::State Lexer::singleQuoteState() {
+	char ch = chars.consume();
+	if (ch == EOF) {
+		// todo: determine behavior
+		throw std::runtime_error("encountered EOF in single quote parsing");
+	}
+	state_data.word.push_back(ch);
+	if (isSingleQuote(ch)) {
+		return State::Word;
+	}
+	return State::SingleQuote;
+}
+
+Lexer::State Lexer::doubleQuoteStateStart() {
+	char ch = chars.consume();
+	D_ASSERT(isDoubleQuote(ch));
+	state_data.word.push_back(ch);
+	return State::DoubleQuote;
+}
+
+Lexer::State Lexer::doubleQuoteState() {
+	char ch = chars.consume();
+	if (ch == EOF) {
+		// todo: determine behavior
+		throw std::runtime_error("encountered EOF in double quote lexing");
+	}
+	if (isBackslash(ch)) {
+		// May not push back backslash in case of new line
+		return State::DoubleQuoteBackslash;
+	}
+	state_data.word.push_back(ch);
+	if (isDoubleQuote(ch)) {
+		return State::Word;
+	} else if (isBackTick(ch)) {
+		throw NotImplementedException();
+	} else if (isDollarSign(ch)) {
+		throw NotImplementedException();
+	}
+	return State::DoubleQuote;
+}
+
+// precondition: backslash is already consumed
+// separate state since nothing is actually removed from the input stream here (and it only works for certain
+// characters)
+Lexer::State Lexer::doubleQuoteBackslashState() {
+	char ch = chars.consume();
+	if (ch == EOF) {
+		// todo: determine behavior
+		throw std::runtime_error("encountered EOF in double quote backslash lexing");
+	}
+	if (isNewline(ch)) {
+		return State::DoubleQuote;
+	}
+	state_data.word.push_back('\\');
+	state_data.word.push_back(ch);
+	return State::DoubleQuote;
+}
+
+Lexer::State Lexer::expansionStartState() {
+	char ch = chars.consume();
+	D_ASSERT(ch == '$');
+	state_data.word.push_back(ch);
+	ch = chars.peek();
+	state_data.word.push_back(ch);
+	switch (ch) {
+		case '{':
+			state_data.word.push_back(chars.consume());
+			return State::ParameterExpansion;
+		case '(':
+			throw NotImplementedException();
+			// return CommandSubstitutionStart;
+			// state_data.word.push_back(chars.consume());
+			// if (chars.peek() == '(') {
+			// 	state_data.word.push_back(chars.consume());
+			// 	return State::ArithmeticExpansion;
+			// } else {
+			// 	return State::CommandSubstitution;
+			// }
+		default:
+			// Reasoning: parameter expansion without special behavior
+			return State::Word;
+	}
+}
+
+Lexer::State Lexer::parameterExpansionState() {
```

Comment:
I thought the only logic that had to be performed here was correctly skip over the characters, treat it as a WORD and then later we expand the WORD, potentially creating multiple tokens?

So in that case the expansion would be handled when we request a token

Reply:
> During token recognition no substitutions shall be actually performed, and the result token shall contain exactly the characters that appear in the input (except for <newline> joining), unmodified, including any embedded or enclosing quotes or substitution operators, between the quote mark and the end of the quoted text.

https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_03
(point 5)

Reply:
Ah actually that's not applicable here, point 4 covers `"` `'` and `\` behavior.

It's point 5 that does apply here:
> The shell shall read sufficient input to determine the end of the unit to be expanded

Reply:
Yes, this is unfinished. It's a draft PR (i.e. not to be merged). All chars should be added to the resulting word